### ISSUE TITLE
(Node) Prompt.text cancelling doesn't seem to be working

### ISSUE THREAD

**Issue by BlueHatbRit** [Microsoft/BotBuilder#1803]
In the Node documentation for [Prompts](https://docs.botframework.com/en-us/node/builder/chat/prompts/#collecting-input) there's a paragraph which says the following.

> Prompts return to the caller an IPromptResult. The users response will be contained in the results.response field and may be null. There are a number of reasons for the response to be null. The built-in prompts let the user cancel an action by saying something like ‘cancel’ or ‘nevermind’ which will result in a null response. Or the user may fail to enter a properly formatted response which can also result in a null response. The exact reason can be determined by examining the ResumeReason returned in result.resumed.

This is in reference to the build in Prompts such as `Prompt.text(session, args)`. However when I type "cancel" or "nevermind" to my bot over Slack during a Prompt.text, whatever text is entered is being returned as `result.response` rather than it being `null` as the documentation suggests.

Is this a bug, bad wording, or me getting it wrong? The offending code is below (but simplified length).

```javascript
function(session, args, next) {
    let repo = builder.EntityRecognizer.findEntity(args.entities, 'github.repo.name');
    let req = session.dialogData.req = {
        repo: repo ? repo.entity : null
    };

    if (!req.repo) {
        builder.Prompts.text(session, 'What repo shall I use?');
    } else {
        next();
    }
},

function(session, result, next) {
    let req = session.dialogData.req;
    if (result.response) {
        req.repo = result.response;
    }

    // It'll be null if the user has cancelled
    if (!req.repo) {
        // Say okay, we're leaving it etc...
        session.endDialog();
    }
}
```

**Comment by BlueHatbRit** [comment]
I've tried this out with `Prompt.number(session, 'question');` just now and it's happening here also. Am I missing some link to LUIS with my entity recogniser?

**Comment by BlueHatbRit** [comment]
Any news on this? I need to demo my bot tomorrow afternoon and without this it's going to be really brittle. The documentation clearly says it's a feature so I must be doing something wrong that's causing it to not work, I'm assuming it's a fairly simple fix I just can't find it for the life of me!

Thanks :)

**Comment by Stevenic** [comment]
I'm sorry for the delay. I hope your demo went ok. You need to add a cancelAction() to thedialogyou want to be cancelable.

**Comment by BlueHatbRit** [comment]
@Stevenic thanks for the reply, the demo was fine in the end and I managed to work around it.

Is there any documentation or example for `cancelAction()`? I can't seem to find it through google or navigating the main documentation.

**Comment by BlueHatbRit** [comment]
I found an example in the main project here for future reference of anyone with the same issue - https://github.com/Microsoft/BotBuilder/blob/master/Node/examples/feature-cancelAction/app.js

Closing this issue.

**Comment by Stevenic** [comment]
There's documentation coming to go along with the updated samples.  Be a little careful with the examples as they're based on 3.5 which is currently in preview mode.  You can install 3.5 using `npm install --save botbuilder@next`

**Comment by BlueHatbRit** [comment]
Oh okay that's useful to know, is there any ETA of when 3.5 might hit out of interest? I'm aware you're waiting for other updates to be released first.

**Comment by Stevenic** [comment]
I would have said last week but its on hold for a couple of reasons. My hope is this week and if not then next.  I'm just working on re-writing all of the docs and rolling in minor bug fixes as they  come up.

**Comment by BlueHatbRit** [comment]
Understood. On a side note I love the project, it's great so far just navigating documentation has been a pain recently, thanks for the project!

**Comment by Stevenic** [comment]
BTW... for cancelAction() the only major new thing is really `confirmPrompt` otherwise it's mostly the same. The big additions for 3.5 (other than a complete re-write of the message routing system) that you're likely to be impacted by are `triggerActions()` and global recognizers which you can install using `bot.recognizer()`.  These two features are likely to change the way you build your bots across the board. That's why I'm re-writing all of the docs, to show you how to build bots with these two new tools.

**Comment by Stevenic** [comment]
The new docs are all scenario focused...  There's a new `Conversation Basics` doc which should teach you how to do 90% of what you need to do in one doc. That's one of the things I'm waiting on... Some internal teams are reviewing that doc. So far the feedback has been positive

**Comment by Stevenic** [comment]
You can look at things like the basics-naturalLanguage example for a preview

**Comment by BlueHatbRit** [comment]
That's really great to know, my bot isn't too big and it's just for a uni project mostly right now so moving it to the new 3.5 structure should be fair straight forward I'd imagine. I'm glad the docs are getting updated in that way, I found myself struggling with the limitations of the old ones and diving into the code a lot to find how things worked.

Is there a 3.5 release issue, pull request, or milestone or anything I can hit Watch on for an update of the release? I had a check but can't seem to find one, if not then I'll just keep my eyes peeled over the next couple of weeks.

I'll go and take a look at that now, my bot is using LUIS so that'll be quite relevant. Cheers for your help and the insight!

**Comment by Stevenic** [comment]
So before 3.5 your actions could triggered only by regular expressions. Now you can add your luis model as a global recognizer to get run for every message from the user. You can then specify which intents trigger which actions. Using the new `triggerAction()` you can essentially say that anytime the user says "order pizza" which returns the "OrderPizzaIntent" from LUIS that it should automatically switch to the 'OrderPizzaDialog'. This replaces any active task on the stack but I give you ways to intercept and change that behavior. You can have your 'OrderPizzaDialog' save the current order off for instance in a way that the user can later return to it by saying "continue order" or something. One of the big complaints we've had is that a user couldn't easily interrupt a task and start a new one which this update addresses.   

As for what to monitor. The release will be [here](https://github.com/Microsoft/BotBuilder/releases) but not sure if you can watch that

**Comment by BlueHatbRit** [comment]
This is absolutely perfect, I was going to wait until 3.5 had released to ask about global LUIS handling. It's been really useful for intents but for custom prompts I've found myself wanting to try and identify a particular entity rather than just accepting something with regex which is quite hard to fit to the users needs.

The triggerAction also looks great, the implementation looks really clean as well. It would be good for a user to instantly move onto any dialog without the need to say things like "cancel" a few times to move on, it's much more natural.

Over the next few days I'll be looking at creating some middleware which identifies if you're talking to the bot explicitly or not. I'm looking to try and put the bot into a mutli-user situation like slack. I'll probably wait until 3.5 is out to continue the project as I'm working on a few other things right now. I was wondering if this is something that has been considered as a feature of if it's something you'd consider to be more suitable as a separate library and so won't be looking at? I've had a search and I can't find it in any documentation so I'm assuming it doesn't already exist but I could be missing it.

**Comment by Stevenic** [comment]
I'd actually did think about something similar at one point. I can't remember what I was calling the feature. Something like "smart listening".  The trick with monitoring ambient message streams is knowing when the bot should respond.  You can wait for the bot to be @mentioned but then how long do you continue to listen to the user.  Most people use timers and such.  It turns out that with BotBuilder you have the dialog stack so you can actually be a little smarter.

If I was to build said middleware I would ignore everything from the stream until I was @mentioned. Then once mentioned I would only route messages to the bot while there's a dialog on the stack. The moment the stack goes empty I'd go back to listening for an @mention.  You can set a flag in `session.privateConversationData` to know if you should route or not.

Developers will need to be a little more religious about calling `endDialog()` or `endConversation()` when they're done listening to the user but that's ok.

**Comment by BlueHatbRit** [comment]
I thought the dialog system might be the way to go as well, it should be quite trivial to implement on a basic level using the @mention. Are there / will there be any docs on best practice for creating middleware functions for the botbuilder?
